### PR TITLE
[75] Add ccache to build process

### PR DESCRIPTION
**Pull request by MartinFlores751**

### CODE REVIEW

File: build_and_copy_packages_to_dir.sh [hunk 1]
```diff
@@ -71,25 +71,26 @@ debug_config="-DCMAKE_BUILD_TYPE=Release"
...
+)
+
 while [ -n "$1" ] ; do
     case "$1" in
         --core-only)             core_only=1;;
         -N|--ninja)              make_program_config="-GNinja";
                                  make_program="ninja";;
         -j|--jobs)               shift; build_jobs=$(($1 + 0));;
         -d|--debug)              debug_config="-DCMAKE_BUILD_TYPE=Debug";;
+        -C|--ccache)             common_cmake_args+=(-DCMAKE_CXX_COMPILER_LAUNCHER=ccache);;
```
Comment (alanking) [inline build_and_copy_packages_to_dir.sh:86]: Please add a note about this option to `usage`

**Review by alanking — COMMENTED** [review]
This is very nice.

Does this require a volume mount on the host machine so that the cache persists? Perhaps we should add a note to the README about this as well?

**Comment by MartinFlores751** [comment]
> Does this require a volume mount on the host machine so that the cache persists? Perhaps we should add a note to the README about this as well?

Yep, the mount location is dependent on what the [CCACHE_DIR](https://github.com/irods/irods_development_environment/pull/73/files#diff-b791fa22bf690eb8195a059c143fe30c7e256b653e27991f85ba1a50ce496ad8R105) environment variable is defined as.

Good idea to add this to the README.

File: irods_core_builder.ubuntu18.Dockerfile [hunk 2]
```diff
@@ -101,6 +102,7 @@ ENV PATH=${clang_path}:$PATH
 
 ENV file_extension="deb"
 ENV package_manager="apt-get"
+ENV CCACHE_DIR="/irods_build_cache"
```
Comment (korydraughn) [inline irods_core_builder.ubuntu18.Dockerfile:105]: Is this environment variable special to ccache?
What happens if `--ccache` is provided but the user forgets to provide a volume mount?
What happens if neither of these options are provided?

File: README.md [hunk 3]
```diff
@@ -72,12 +72,26 @@ Available options:
...
+
+```
+$ docker run --rm \
+             -v /full/path/to/irods_repository_clone:/irods_source:ro \
+             -v /full/path/to/irods_build_output_dir:/irods_build \
+             -v /full/path/to/icommands_repository_clone:/icommands_source:ro \
+             -v /full/path/to/icommands_build_output_dir:/icommands_build \
+             -v /full/path/to/packages_output_dir:/irods_packages \
+             -v /full/path/to/build_cache_dir:/irods_build_cache \
+             irods-core-builder-ubuntu16 --ccache
```
Comment (trel) [inline README.md:92]: `--ccache` at the end?   or a line above... before the image name?

File: README.md [hunk 4]
```diff
@@ -72,12 +72,26 @@ Available options:
     -d, --debug             Build with symbols for debugging
     -j, --jobs              Number of jobs for make tool
     -N, --ninja             Use ninja builder as the make tool
+    -C, --ccache            Enables ccache for rapid subsequent builds
```
Comment (trel) [inline README.md:75]: is alpha order a thing we want?   both here and in `build_and_copy_packages_to_dir.sh`...

File: README.md (hunk 3 above)
Comment (korydraughn) [inline README.md:92]: Consider highlighting the things of interest here.

As it is, the reader has to scan the command to see what's different from the previous docker run example.

**Review by korydraughn — COMMENTED** [review]
Very close.

File: README.md (hunk 3 above)
Comment (MartinFlores751) [reply README.md:92]: At the end. It should be like the other options, example: `irods-core-builder-ubuntu16 -j42 --ccache --ninja`. Lines above are for docker and not passed to the script.

File: README.md (hunk 4 above)
Comment (MartinFlores751) [reply README.md:75]: Didn't notice the ordering… I'll make it pretty again.

File: irods_core_builder.ubuntu18.Dockerfile (hunk 2 above)
Comment (MartinFlores751) [reply irods_core_builder.ubuntu18.Dockerfile:105]: Yep, that is an environment variable special to ccache.

If both mount and option are forgotten, ccache is not used.
If only the mount is provided, ccache is not used.
If only the `--ccache` option is provided, you make your build slower.
If both options are provided, everything is happy.

Not sure what's the best way to handle the case where only `--ccache` is provided.

Comment (korydraughn) [reply irods_core_builder.ubuntu18.Dockerfile:105]: Does ccache create /irods_build_cache in the container? I'm guessing it does?

Comment (MartinFlores751) [reply irods_core_builder.ubuntu18.Dockerfile:105]: Just checked. It seems to be creating and using the `CCACHE_DIR` in the case where you provide the `--ccache` flag but don't provide a mount.

Comment (trel) [reply irods_core_builder.ubuntu18.Dockerfile:105]: so there's a default value for the `CCACHE_DIR`?  what is that?

Comment (MartinFlores751) [reply irods_core_builder.ubuntu18.Dockerfile:105]: There is a default, being the `$HOME/.ccache` I believe, but here we set the `CCACHE_DIR` to always be `/irods_build_cache`.

Comment (trel) [reply irods_core_builder.ubuntu18.Dockerfile:105]: very good.   thanks.

**Comment by korydraughn** [comment]
What's the status of this PR?

**Comment by MartinFlores751** [comment]
Need to add ccache to other Dockerfiles (i.e., Ubuntu 20, CentOS 7, etc.) is about it.

Think there was a thought of using volume mounts instead of bind mounts for the ccache directory? Could also try that if it sounds interesting, to avoid the case of cache + no bind mount provided.

**Comment by korydraughn** [comment]
It looks like we're already using volume mounts (as documented in the README changes).

How would a bind mount work?
Is that easy to use?
Are bind mounts a commonly used feature?

**Comment by MartinFlores751** [comment]
I think that's a bit of a misuse of the word 'volume' in the context of docker on our part. [Bind mounts](https://docs.docker.com/storage/bind-mounts/) seem like what we're already doing. I might be wrong by using 'mounts', but [volumes](https://docs.docker.com/storage/volumes/) are docker-specific persistent storage.

Volumes seem to work similarly, except that you can only access/share them with docker containers(?), and that, based on how a container is configured, may not explicitly need a 'mount' specified like bind mount's `-v`. This may let us avoid the cache & trash situation, in my understanding.

It seems easy to use. The primary downside being that if we put the cache in a volume, we can't use the generated cache on the host without going through some extra steps to extract/copy the data in the volume to the host.

**Comment by korydraughn** [comment]
> The primary downside being that if we put the cache in a volume, we can't use the generated cache on the host without going through some extra steps to extract/copy the data in the volume to the host.

I think that is a good enough reason to stick to using `-v`. That hasn't caused any issues so far and should be familiar to docker users.

As long as we document things, we're okay. And if someone forgets to specify the `-v`, then they'll feel that immediately.

File: README.md [hunk 5]
```diff
@@ -78,6 +79,15 @@ Available options:
 ```
 The `--custom-externals` option allows you to specify a location in the builder container where there are built iRODS externals. This is useful if the platform does not have built externals available in the RENCI repositories or when trying out a modification to the externals. To do this, use a volume mount when running the docker container and specify the mount name with this option (i.e. the path inside the container).
 
+The `--ccache` option allows you to utilize `ccache` in the build process, speeding up subsequent builds of icommands and the iRODS server. Using this option requires the use of an additional volume mount to the `/irods_build_cache` directory, an abbreviated example is as follows:
+
```
Comment (korydraughn) [inline README.md:83]: Let's remove this empty line for consistency.

**Review by korydraughn — COMMENTED** [review]
Just a couple things left to do:
- Update the README 
- Create a new issue and include it in the commit message

Please leave off the pound until the PR is approved.

**Review by korydraughn — APPROVED** [review]

**Review by alanking — APPROVED** [review]
I think it's ready!

**Comment by korydraughn** [comment]
Let's add the pound and get it merged.

**Pull request merged**